File: src/main/java/org/modelingvalue/nelumbo/AstElement.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2025 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo;

import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Map;
import org.modelingvalue.nelumbo.patterns.Functor;
import org.modelingvalue.nelumbo.syntax.Token;

public interface AstElement {

    Token firstToken();

    Token lastToken();

    static Token firstToken(List<AstElement> elements) {
        for (AstElement element : elements) {
            Token first = element.firstToken();
            if (first != null) {
                return first;
            }
        }
        return null;
    }

    static Token lastToken(List<AstElement> elements) {
        for (AstElement element : elements.reverse()) {
            Token last = element.lastToken();
            if (last != null) {
                return last;
            }
        }
        return null;
    }

    boolean isMeta();

    List<Integer> getBranches(Functor functor);

    void setBranches(Map<Functor, List<Integer>> branches);

    void setCycleDepth(int depth);

    int getCycleDepth();
}

File: web-ts/src/AstElement.ts
/**
 * AstElement interface - common interface for Token and Node.
 * @JAVA_REF org.modelingvalue.nelumbo.AstElement
 */

import { List } from 'immutable';
import { Token } from './syntax/Token';

/**
 * Interface for AST elements (Token and Node).
 */
export interface AstElement {
  /**
   * Get the first token of this element.
   */
  firstToken(): Token | null;

  /**
   * Get the last token of this element.
   */
  lastToken(): Token | null;

  /**
   * Deparse this element to a string buffer.
   */
  deparse(sb: string[]): void;
}

/**
 * Utility functions for AstElement.
 */
export const AstElementUtil = {
  /**
   * Get the first token from a list of elements.
   */
  firstToken(elements: List<AstElement>): Token | null {
    for (const element of elements) {
      const first = element.firstToken();
      if (first !== null) {
        return first;
      }
    }
    return null;
  },

  /**
   * Get the last token from a list of elements.
   */
  lastToken(elements: List<AstElement>): Token | null {
    for (const element of elements.reverse()) {
      const last = element.lastToken();
      if (last !== null) {
        return last;
      }
    }
    return null;
  },
};
